zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -284,17 +286,30 @@ impl DbFormatChange {
                 "The Sapling note commitment tree for the genesis block should be in the database.",
             );
             let mut last_height = Height(1);
+
+            // We use this dummy cap in the loop below. It resolves an edge case when the pruning
+            // reaches the `initial_tip_height`.
+            let dummy_cap = Some((
+                initial_tip_height.next(),
+                Arc::new(sapling::tree::NoteCommitmentTree::default()),
+            ))
+            .into_iter();
+
             // Run through all the trees in the finalized chain.
-            for (height, tree) in db.sapling_tree_by_height_range(Height(1)..=initial_tip_height) {
+            for (height, tree) in db
+                .sapling_tree_by_height_range(Height(1)..=initial_tip_height)
+                .chain(dummy_cap)
+            {
```

Comment:
This looks like a good fix, but if the state hasn't reached the activation height yet, the `initial_tip_height` will remain in the database, because the `dummy_cap` tree is equal to the `initial_tip_height` tree. So neither of these conditions is satisfied:
`last_tree != tree || height == initial_tip_height`

I'll push a suggested refactor to another PR.

Reply:
Once we fix the versioning in PR #7385, this bug would have been caught by the checkpoint test, because the mandatory checkpoint is before NU5 activation

Reply:
PR #7391 should fix this, I tested it on my legacy state with the fix in PR #7385.